### PR TITLE
SLE-10768: Work on ha-cluster-preflight-check

### PR DESCRIPTION
**Pull request by tomschr**
### Description

* Add entities for `test.script`/`cmd.test.script`
* Split section "sec-ha-inst-quick-test" into two sect2's.

See https://jira.suse.com/browse/SLE-10768

### Checklist
* Check all items that apply.

*Are backports required?* needs to be checked

- [ ] To maintenance/SLEHA12SP4
- [ ] To maintenance/SLEHA15
- [ ] To maintenance/SLEHA15SP1

### CODE REVIEW

**Comment by tomschr** [comment]
@liangxin1300 Hi Xin, would you be so kind of reviewing the small section? It belongs to the Quick Installation Start guide and describes your script.

----

Screenshot_20200320_152230

**Comment by liangxin1300** [comment]
> @liangxin1300 Hi Xin, would you be so kind of reviewing the small section? It belongs to the Quick Installation Start guide and describes your script.
> 
> I'm having some minor problems of how to interpret the output of `ha-cluster-preflight-check -e`. I would have expected that the warning or info messages contain also the nodes. At the moment, I can only see _which_ service failed, but not _where_. Is that true? Bug/feature? 

@tomschr , Thanks for the documents! Great job!
For now, `-e` option is just for `local` environment. Maybe it's not accurate that the help does not show that, maybe we should change it as
```
# ha-cluster-preflight-check -h
-e, --env-check         Check local environment
```

> Thanks Xin, much appreciated!
> 
> Screenshot_20200320_152230

**Comment by tomschr** [comment]
> For now, `-e` option is just for `local` environment. Maybe it's not accurate that the help does not show that, maybe we should change it as

@liangxin1300 
Ahh, that makes sense! Thanks Xin for the explanation. I will add that to the explanation.

Any other things in the doc that is wrong, ambiguous, or maybe phrased unfortunately? :wink:

**Comment by liangxin1300** [comment]
@tomschr There was also an option `--split-brain-iptables` in this package, help message is:
```
--split-brain-iptables  Make split brain by blocking corosync ports
```
This option is to simulate split brain, and to check whether one node can be fenced as expected

**Comment by tomschr** [comment]
> There was also an option `--split-brain-iptables` in this package, help message is:

Ahh, right! This is missing in the list. Will add it.


> And for each --kill-* testcase, a report will be created at directory: /var/lib/ha-cluster-preflight-check
> Report will includes test case description, action logging and explanation for possible results.

Not sure what you want to say here. Was it just additional information or something I need to change? I've described the `--kill*` options already. :wink:

**Comment by liangxin1300** [comment]
> > There was also an option `--split-brain-iptables` in this package, help message is:
> 
> Ahh, right! This is missing in the list. Will add it.
> 
> > And for each --kill-* testcase, a report will be created at directory: /var/lib/ha-cluster-preflight-check
> > Report will includes test case description, action logging and explanation for possible results.
> 
> Not sure what you want to say here. Was it just additional information or something I need to change? I've described the `--kill*` options already.

It's an additional information. You can append it to the last line

**Comment by tomschr** [comment]
> > Not sure what you want to say here. Was it just additional information or something I need to change? I've described the `--kill*` options already.
> 
> It's an additional information. You can append it to the last line

Sorry, I'm confused. Currently, I have this:

```xml
<formalpara>
       <title>Killing Test Cases <option>-kill-sbd</option>/<option>-kill-corosync</option>/<option>-kill-pacemakerd</option></title>
       <para>
        Kill the SBD, &corosync;, and &pace; daemons.
        After running such a test, you can find a report in
        <filename>/var/lib/ha-cluster-preflight-check</filename>.
        The report includes test case description, action logging,
        and explains possible results.
       </para>
</formalpara>
```

you have this:

```
And for each --kill-* testcase, a report will be created at directory: /var/lib/ha-cluster-preflight-check
Report will includes test case description, action logging and explanation for possible results.
```

Apart from the wording, isn't it basically the same?

**Comment by liangxin1300** [comment]
> > > Not sure what you want to say here. Was it just additional information or something I need to change? I've described the `--kill*` options already.
> > 
> > 
> > It's an additional information. You can append it to the last line
> 
> Sorry, I'm confused. Currently, I have this:
> 
> ```
> <formalpara>
>        <title>Killing Test Cases <option>-kill-sbd</option>/<option>-kill-corosync</option>/<option>-kill-pacemakerd</option></title>
>        <para>
>         Kill the SBD, &corosync;, and &pace; daemons.
>         After running such a test, you can find a report in
>         <filename>/var/lib/ha-cluster-preflight-check</filename>.
>         The report includes test case description, action logging,
>         and explains possible results.
>        </para>
> </formalpara>
> ```
> 

Sorry, I didn't noticed that you already mentioned it. That's the same meanings!

Thank you @tomschr !

**Comment by tomschr** [comment]
> Sorry, I didn't noticed that you already mentioned it. That's the same meanings!

@liangxin1300 Thanks Xin! :+1: Now it's clear. :wink: Thanks for the clarification. :+1:

**Comment by tomschr** [comment]
Updated section

---

Screenshot_20200323_154253

File: xml/art_sle_ha_install_quick.xml [hunk 1]
```diff
@@ -715,15 +715,20 @@ softdog                16384  1</screen>
...
-    <remark>toms 2016-07-27: Fate#321073
+   <sect2 xml:id="sec-ha-inst-quick-test-resource-failover">
+    <title>Testing Resource Failover</title>
+    <para>
+     As a quick test, the following procedure checks on resource failovers:
+    </para>
+   <remark>toms 2016-07-27: Fate#321073
     Tool for Standardize Testing of Basic Cluster Functionality</remark>
    <procedure xml:id="pro-ha-inst-quick-test">
-    <title>Testing Resource Failover</title>
```
Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml]: Removing the ID from the procedure breaks the PDF output on p.12: 

```Procedure 4, "" is a simple test to check...```

Please reference the new section ID instead.

File: xml/art_sle_ha_install_quick.xml [hunk 2]
```diff
@@ -715,15 +715,20 @@ softdog                16384  1</screen>
...
+    script.
     <!-- toms 2016-08-01: For SP3, refer to the "Testing Cluster Guide" or
      to Lars Pinnes iptable rules
     -->
    </para>
-    <remark>toms 2016-07-27: Fate#321073
+   <sect2 xml:id="sec-ha-inst-quick-test-resource-failover">
+    <title>Testing Resource Failover</title>
+    <para>
+     As a quick test, the following procedure checks on resource failovers:
```
Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml]: The procedure is not run automatically, but requires human interaction, therefore better phrase in active voice:
' To check resource failover, execute the following steps:' (or similar)

File: xml/art_sle_ha_install_quick.xml [hunk 3]
```diff
@@ -788,6 +793,98 @@ softdog                16384  1</screen>
        <keycap function="control"/><keycap>C</keycap>
       </keycombo>.
    </para>
+   </sect2>
+
+   <sect2 xml:id="sec-ha-inst-quick-test-with-cluster-script">
+    <title>Testing with the &test.script; Script</title>
+    <para>
+     The command &cmd.test.script; is a tool to run standardized tests
```
Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml]: 'script', 'command', 'tool' - why not use the same term for the same thing?

File: xml/art_sle_ha_install_quick.xml [hunk 4]
```diff
@@ -788,6 +793,98 @@ softdog                16384  1</screen>
...
+     for a cluster.
+     It triggers cluster failures and verify configuration to find
```
Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml]: verify -> verifies the cluster configuration

File: xml/art_sle_ha_install_quick.xml [hunk 5]
```diff
@@ -788,6 +793,98 @@ softdog                16384  1</screen>
...
+    </para>
+    <para>
+     The script supports the following checks:
+    </para>
+    <itemizedlist>
+     <listitem>
+      <formalpara>
+       <title>Environment Check <option>-e</option>/<option>--env-check</option></title>
+       <para>
+        Check if hostname is resolvable, time service, watchdog, and
```
Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml]: Use the same grammatical structure for the explanation in the list, I suggest to use third person singular:
Check -> Checks if...

Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml]: hostname -> host names (spelling according to our style guide) and I would use the plural as the script probably checks more than one host name...

Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml]: This is very much reduced to keywords - how about phrasing it in a complete sentence: 
Checks if all host names can be resolved, and if time synchronization, watchdog, and firewall settings are configured correctly.

File: xml/art_sle_ha_install_quick.xml [hunk 6]
```diff
@@ -788,6 +793,98 @@ softdog                16384  1</screen>
...
+       firewall.
+       </para>
+      </formalpara>
+     </listitem>
+     <listitem>
+      <formalpara>
+       <title>Cluster State Check <option>-c</option>/<option>--cluster-check</option></title>
+       <para>
+        Check cluster service, &stonith;/fence configuration, nodes,
```
Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml]: Same here: Please elaborate a bit more.

File: xml/art_sle_ha_install_quick.xml [hunk 7]
```diff
@@ -788,6 +793,98 @@ softdog                16384  1</screen>
...
+        and resources.
+       </para>
+      </formalpara>
+     </listitem>
+     <listitem>
+      <formalpara>
+       <title>Split Brain Check <option>--split-brain-iptables</option></title>
+       <para>
+        Simulates split brain scenario by blocking the &corosync; port.
```
Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml]: Simulates a split brain scenario...

File: xml/art_sle_ha_install_quick.xml [hunk 8]
```diff
@@ -788,6 +793,98 @@ softdog                16384  1</screen>
...
+        Checks wheather one node can be fenced as expected.
+       </para>
+      </formalpara>
+     </listitem>
+     <listitem>
+      <formalpara>
+       <title>Killing Test Cases <option>-kill-sbd</option>/<option>-kill-corosync</option>/<option>-kill-pacemakerd</option></title>
```
Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml]: Probably too late to change it in the software for SLE HA 15 SP2, but 'killing' is a rather unfortunate term to use in this regard. At least in the documentation, I would change it to something like 
```Stops the daemons for SBD, &corosync;, and &pace;.```

File: xml/art_sle_ha_install_quick.xml [hunk 9]
```diff
@@ -788,6 +793,98 @@ softdog                16384  1</screen>
...
+        <filename>/var/lib/ha-cluster-preflight-check</filename>.
+        The report includes test case description, action logging,
+        and explains possible results.
+       </para>
+      </formalpara>
+     </listitem>
+     <listitem>
+      <formalpara>
+       <title>Fence Node Check <option>--fence-node</option></title>
+       <para>Fence a specific node</para>
```
Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml]: Fence -> Fences a specific node (or even: 'Fences the specified node' if one can specify a node with the option). Best check the options in a test system.

File: xml/art_sle_ha_install_quick.xml (hunk 1 above)
Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml]: Argh, right. Strange, this case should be caught by the stylesheets. I refered to the section as you advised.

File: xml/art_sle_ha_install_quick.xml (hunk 2 above)
Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml]: Ok, done.

File: xml/art_sle_ha_install_quick.xml (hunk 3 above)
Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml]: Argh, right. Ok, try to be consistent.

File: xml/art_sle_ha_install_quick.xml (hunk 4 above)
Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml]: Fixed.

File: xml/art_sle_ha_install_quick.xml (hunk 5 above)
Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml]: Ok, fixed.

Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml]: Right, fixed.

Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml]: Ok, fixed.

File: xml/art_sle_ha_install_quick.xml (hunk 6 above)
Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml]: @liangxin1300 I've tried to read the source code, but I guess, it's easier to ask you directly. :wink: 

Could you elaborate a bit?

File: xml/art_sle_ha_install_quick.xml (hunk 9 above)
Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml]: The command line option is: `--fence-node NODE       Fence specific node`. I changed the line to:

    Fences the specific node passed from the command line.

File: xml/art_sle_ha_install_quick.xml (hunk 8 above)
Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml]: @liangxin1300 Xin, is it too late to change the options in the command? Could we use `--stop-*` instead?  I can also open an issue in your repo, if it's possible. Just asking. :wink:

Nevertheless, I've changed the line as Tanja advised.

Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml]: Ok, clarified difference between "stopping" and "killing" of resources in an email. Thanks to Lars. Setting question as resolved.

File: xml/art_sle_ha_install_quick.xml (hunk 6 above)
Comment (liangxin1300) [reply xml/art_sle_ha_install_quick.xml]: @tomschr @taroth21 
More descriptions can be:
* Check cluster service
   Check whether cluster service(pacemaker/corosync) is enabled and running;
* Check stonith
  Check whether stonith is enabled;
  Check whether stonith related resources is configured and started;
  If we configured sbd, check whether sbd service is started
* Check nodes
  Show current DC node
  Check whether cluster have quorum
  Show Online/OFFLINE/UNCLEAN nodes if cluster has
* Check resources
  Show number of Started/Stopped/FAILED resources
```

Comment (liangxin1300) [reply xml/art_sle_ha_install_quick.xml]: And for environment check, that is, `-e/--env-check`, more details can be:
#### Check the local node cluster related environment
* Check hostname resolvable
* Check whether time service is enabled and started
* Check whether this node configured watchdog
* Check whether firewalld service is started and whether cluster related ports are opened

Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml]: Thanks @liangxin1300! Much appreciated!

I've added it as a list of questions.

**Comment by tomschr** [comment]
@taroth21 If your time permits, could you look at the improved list? I'm a bit unsure if this would be useful. I used a list with questions.

I will post an image soon.

File: xml/art_sle_ha_install_quick.xml [hunk 10]
```diff
@@ -788,6 +794,139 @@ softdog                16384  1</screen>
...
+        </para>
+       </listitem>
+       <listitem>
+        <para>
+         Has the current node configured watchdog?
+        </para>
+       </listitem>
+       <listitem>
+        <para>
+         Is the firewalld service started and are cluster related ports
```
Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml:835]: cluster-related?

File: xml/art_sle_ha_install_quick.xml [hunk 11]
```diff
@@ -788,6 +794,139 @@ softdog                16384  1</screen>
...
+      </formalpara>
+      <itemizedlist>
+       <listitem>
+        <para>
+         Are cluster services (&pace;/&corosync;) enabled and running?
+        </para>
+       </listitem>
+       <listitem>
+        <para>
+         Is &stonith; enabled? It also checks, whether &stonith; related
```
Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml:854]: &stonith;-related

File: xml/art_sle_ha_install_quick.xml [hunk 12]
```diff
@@ -788,6 +794,139 @@ softdog                16384  1</screen>
...
+         resources are configured and started.
+         If you configured SBD, is the SBD service started?
+        </para>
+       </listitem>
+       <listitem>
+        <para>
+         Have clusters quorum? Shows current DC nodes and nodes which are online,
```
Comment (taroth21) [inline xml/art_sle_ha_install_quick.xml:861]: Do clusters have quorum? (better check with Liam, too).

Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml:861]: @lproven what do you think? Should I better write `Do cluster nodes have quorum?` Or maybe there is even a better alternative? Thanks!

Comment (lproven) [reply xml/art_sle_ha_install_quick.xml:861]: I suggest:
> Does the cluster have a quorum?

File: xml/art_sle_ha_install_quick.xml (hunk 11 above)
Comment (lproven) [reply xml/art_sle_ha_install_quick.xml]: Also remove the comma after "checks".

File: xml/art_sle_ha_install_quick.xml [hunk 13]
```diff
@@ -788,6 +794,139 @@ softdog                16384  1</screen>
...
+        <para> Are host names resolvable? </para>
+       </listitem>
+       <listitem>
+        <para>
+         Is the time service enabled and started?
+        </para>
+       </listitem>
+       <listitem>
+        <para>
+         Has the current node configured watchdog?
```
Comment (lproven) [inline xml/art_sle_ha_install_quick.xml:830]: I need to read the whole doc for context but I think I would suggest:
"Does the current node have a watchdog configured?"

File: xml/art_sle_ha_install_quick.xml [hunk 14]
```diff
@@ -788,6 +794,139 @@ softdog                16384  1</screen>
...
+        </para>
+       </listitem>
+       <listitem>
+        <para>
+         Is the firewalld service started and are cluster-related ports
```
Comment (lproven) [inline xml/art_sle_ha_install_quick.xml:835]:
```suggestion
         Is the <command>firewalld</command> service started and are cluster-related ports
```

File: xml/art_sle_ha_install_quick.xml (hunk 12 above)
Comment (lproven) [reply xml/art_sle_ha_install_quick.xml:861]: "Does the cluster have a quorum?"
or
"Is there a quorum of cluster nodes?"

File: xml/art_sle_ha_install_quick.xml (hunk 13 above)
Comment (tomschr) [reply xml/art_sle_ha_install_quick.xml:830]: Yes, that sounds better. Used your suggestion.

**Comment by tomschr** [comment]
Thanks @lproven much appreciated!

**Pull request merged**